File: src/frontend/src/types/lifecycle.ts
/**
 * Shared lifecycle types used across all entity types.
 *
 * EntityStatus replaces DataProductStatus, ContractStatus, DatasetStatus, AssetStatus.
 * PublicationScope replaces the boolean `published` flag.
 */

export type EntityStatus =
  | 'draft'
  | 'sandbox'
  | 'proposed'
  | 'under_review'
  | 'approved'
  | 'active'
  | 'deprecated'
  | 'retired';

export type PublicationScope = 'none' | 'domain' | 'organization' | 'external';

export interface CertificationInfo {
  certification_level: number | null;
  inherited_certification_level: number | null;
  certified_at: string | null;
  certified_by: string | null;
  certification_expires_at: string | null;
  certification_notes: string | null;
}

export interface PublicationInfo {
  publication_scope: PublicationScope;
  published_at: string | null;
  published_by: string | null;
}

export interface CertificationLevel {
  id: string;
  level_order: number;
  name: string;
  description: string | null;
  icon: string | null;
  color: string | null;
}

export const ENTITY_STATUS_LABELS: Record<EntityStatus, string> = {
  draft: 'Draft',
  sandbox: 'Sandbox',
  proposed: 'Proposed',
  under_review: 'Under Review',
  approved: 'Approved',
  active: 'Active',
  deprecated: 'Deprecated',
  retired: 'Retired',
};

export const ENTITY_STATUS_COLORS: Record<EntityStatus, string> = {
  draft: 'bg-gray-100 text-gray-800 dark:bg-gray-700 dark:text-gray-300',
  sandbox: 'bg-sky-100 text-sky-800 dark:bg-sky-900 dark:text-sky-300',
  proposed: 'bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-300',
  under_review: 'bg-amber-100 text-amber-800 dark:bg-amber-900 dark:text-amber-300',
  approved: 'bg-indigo-100 text-indigo-800 dark:bg-indigo-900 dark:text-indigo-300',
  active: 'bg-green-100 text-green-800 dark:bg-green-900 dark:text-green-300',
  deprecated: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900 dark:text-yellow-300',
  retired: 'bg-red-100 text-red-800 dark:bg-red-900 dark:text-red-300',
};

export const PUBLICATION_SCOPE_LABELS: Record<PublicationScope, string> = {
  none: 'Not Published',
  domain: 'Domain',
  organization: 'Organization',
  external: 'External',
};

File: src/backend/src/models/lifecycle.py
"""Shared lifecycle enums and models used across all entity types.

Provides:
- EntityStatus: unified lifecycle status enum (replaces DataProductStatus, ContractStatus, etc.)
- PublicationScope: scoped publication visibility
- CertificationInfo / PublicationInfo: embedded metadata models
- VALID_TRANSITIONS: per-entity-type transition maps
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class EntityStatus(str, Enum):
    """Unified lifecycle status for all governed entities."""
    DRAFT = "draft"
    SANDBOX = "sandbox"
    PROPOSED = "proposed"
    UNDER_REVIEW = "under_review"
    APPROVED = "approved"
    ACTIVE = "active"
    DEPRECATED = "deprecated"
    RETIRED = "retired"


class PublicationScope(str, Enum):
    """Visibility scope when an entity is published."""
    NONE = "none"
    DOMAIN = "domain"
    ORGANIZATION = "organization"
    EXTERNAL = "external"


# Per-entity-type transition maps (status -> list of allowed next statuses)
DATA_PRODUCT_TRANSITIONS: dict[str, list[str]] = {
    "draft": ["sandbox", "proposed", "deprecated"],
    "sandbox": ["draft", "proposed", "deprecated"],
    "proposed": ["draft", "under_review", "deprecated"],
    "under_review": ["draft", "approved", "deprecated"],
    "approved": ["active", "draft", "deprecated"],
    "active": ["deprecated"],
    "deprecated": ["retired", "active"],
    "retired": [],
}

DATA_CONTRACT_TRANSITIONS: dict[str, list[str]] = {
    "draft": ["proposed", "deprecated"],
    # A steward can approve directly from "proposed" (the approve endpoint is
    # documented as PROPOSED/UNDER_REVIEW -> APPROVED). "under_review" stays an
    # optional intermediate step rather than a mandatory one.
    "proposed": ["draft", "under_review", "approved", "deprecated"],
    "under_review": ["draft", "approved", "deprecated"],
    "approved": ["active", "draft", "deprecated"],
    "active": ["deprecated"],
    "deprecated": ["retired", "active"],
    "retired": [],
}

DATASET_TRANSITIONS: dict[str, list[str]] = {
    "draft": ["active", "deprecated"],
    "active": ["deprecated"],
    "deprecated": ["retired", "active"],
    "retired": [],
}

ASSET_TRANSITIONS: dict[str, list[str]] = {
    "draft": ["active", "deprecated"],
    "active": ["deprecated"],
    "deprecated": ["retired", "active"],
    "retired": [],
}


class CertificationInfo(BaseModel):
    """Certification metadata embedded in entity responses."""
    certification_level: Optional[int] = Field(None, description="Ordinal of the certification level")
    inherited_certification_level: Optional[int] = Field(None, description="Inherited via relationships")
    certified_at: Optional[datetime] = None
    certified_by: Optional[str] = None
    certification_expires_at: Optional[datetime] = None
    certification_notes: Optional[str] = None

    model_config = {"from_attributes": True}


class PublicationInfo(BaseModel):
    """Publication metadata embedded in entity responses."""
    publication_scope: PublicationScope = PublicationScope.NONE
    published_at: Optional[datetime] = None
    published_by: Optional[str] = None

    model_config = {"from_attributes": True}
